Add new ColorView & ColorPicker Controls

robloo opened the pull request:
## What does the pull request do?

Adds the last two `ColorView` and `ColorPicker` controls (in the Avalonia.Controls namespace). These are stand-alone controls not based on WinUI code. This closes https://github.com/AvaloniaUI/Avalonia/issues/4179.

These are the final fourth and fifth major controls towards implementing the ColorPicker.

## What is the current behavior?

Only the following primitives are implemented for color-related controls:

 1. `ColorSpectrum` https://github.com/AvaloniaUI/Avalonia/issues/5743
    - PR https://github.com/AvaloniaUI/Avalonia/pull/7842
 2. `ColorSlider`
    - PR https://github.com/AvaloniaUI/Avalonia/pull/8050
 3. `ColorPresenter`
    - PR https://github.com/AvaloniaUI/Avalonia/pull/8050

## What is the updated/expected behavior with this PR?

Adds the two remaining color-related controls completing the `ColorPicker` functionality.

### ColorView

This is the 'canvas' version of the control that does not have a drop down button. ColorPicker hosts this in a drop down button flyout.

| Spectrum | Palette | Components |
|--|--|--|
| ![image](https://user-images.githubusercontent.com/17993847/177462548-a6985ba3-e920-4fed-be51-3c92e113e279.png) | ![image](https://user-images.githubusercontent.com/17993847/177462605-a1357788-91f7-4202-98fb-2fc8a722b4c5.png) | ![image](https://user-images.githubusercontent.com/17993847/177462730-e11a56fa-b651-414e-b6af-42ba1e43ee65.png) |

### ColorPicker

This is the main control to use for color selection and editing and appears as a drop down button.

https://user-images.githubusercontent.com/17993847/177463139-8343e735-a9e3-4975-8293-a600ccad7843.mp4

## How was the solution implemented (if it's not obvious)?

This is a brand-new implementation and attempts to fix past wrongs with all past color pickers starting with the name.

 * The WinUI ColorPicker isn't a picker at all (compared to other pickers) so this implementation clearly separates the `ColorView` from the actual `ColorPicker` implemented in a drop-down.
 * This version of the control attempts to do everything it can in the XAML control template keeping code-behind to an absolute minimum. This significantly increases composability and enables app developers to customize every part of these controls (and even the primitives in most cases).
 * Primitives such as the ColorSlider and ColorSpectrum can still be used separately enabling app developers to create separate color picker implementations.
 * `HsvColor` is heavily used in all controls and was also implemented into base Avalonia alongside `Color` and `HslColor`. This simplified code-behind and also made binding of color between primitives and controls a lot more straightforward.
 * `HsvColor` along with `ColorSlider` together unlock a lot of power with this control compared to the one in WinUI (and enable it to be easily re-templated).
 * Many new properties (more than in WinUI) were added to control all aspects of the ColorView visibility. Each tab can be separately hidden along with most individual subsections. This allows a lot of design customization without having to re-template or drill into styles.
 * New properties such as `SelectedIndex` and `ColorModel` allow customizing the control putting it in a pre-defined state. For example: the WinUI ColorPicker always defaults to RGB and this cannot be changed in code or XAML. This implementation does not have such limitations.

## Checklist

- [ ] ~Added unit tests (if possible)?~
- [x] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Documentation with user documentation
       - Will be done later

## Breaking changes

None

## Obsoletions / Deprecations

None

## Fixed issues

Closes https://github.com/AvaloniaUI/Avalonia/issues/4179

robloo committed:
Add IColorPalette

robloo committed:
Add initial ColorView.Properties

robloo committed:
Add ColorViewTab enum

robloo committed:
Add FluentColorPalette

robloo committed:
Fix ColorToHexConverter

robloo committed:
Add incomplete ColorView control template

robloo committed:
Add braces

robloo committed:
Fix index clamping

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0020723-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

robloo committed:
Further work on ColorView

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0020724-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

robloo committed:
Add new EnumToBooleanConverter

robloo committed:
Implement ColorModel switching in ColorView

robloo committed:
Improve comments

robloo committed:
Set MaxHue of ColorSlider to the same value as ColorSpectrum

robloo committed:
Remove EnumValueEqualsConverter replaced by EnumToBooleanConverter

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0020731-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

robloo committed:
Remove IsAutoUpdatingEnabled property from ColorSlider

This was unused and is an unnecessary complexity

robloo committed:
Finish ColorView design/style

robloo committed:
Implement the hex input TextBox in ColorView

robloo commented on src/Avalonia.Controls.ColorPicker/Themes/Fluent/ColorView.xaml:
```diff
@@ -0,0 +1,562 @@
+                                          Grid.Row="0"
+                                          Grid.RowSpan="2"
+                                          Components="{TemplateBinding ColorSpectrumComponents}"
+                                          HsvColor="{Binding $parent[ColorView].HsvColor}"
```
@maxkatz6 I have a question about binding here. `{TemplateBinding HsvColor}` doesn't seem to work in this case so I've had to replace it with `{Binding $parent[ColorView].HsvColor}`. This applies several places. Do you know why that might be? This really should be fixed so normal TemplateBinding can be used.

robloo replied:
This is because TemplateBinding is only possible OneWay (I haven't thought about that in a while, I guess it just rarely comes up for me). I'll just use a TemplatedParent relative source then. The assumption is all of this works the same as WPF.

timunie replied:
I thought you can have `Components="{TemplateBinding ColorSpectrumComponents, Mode=TwoWay}"` but iirc you need to explicitly set it.

robloo replied:
I had it specified as TwoWay at one point. I can test again. However, if this works like WPF it isnt possible in a TemplateBinding to have TwoWay mode.

timunie replied:
Avalonia is better than WPF here imo.

robloo replied:
Well, I do think I tested that specifically. Could have made a mistake though. I was changing a few things at once to figure our what the issue was.

robloo replied:
@timunie I tested this again and confirmed it doesn't work. It behaves the same as WPF. You have to use the full binding syntax for TwoWay.

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0020732-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

robloo committed:
Implement ColorSlider component value rounding

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0020733-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

robloo committed:
Add initial ColorPicker control

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0020734-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

robloo committed:
Complete initial ColorPicker

robloo committed:
Implement IsAlphaVisible and IsComponentTextInputVisible

robloo committed:
Implement IsHexInputVisible

robloo committed:
Merge branch 'master' into colorpicker2

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0020792-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

robloo committed:
Implement remaining ColorView properties

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0020798-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

robloo commented:
@timunie 

>>Is it a good choice to have an enum for the tabs? I think one may want to extend the control later on with a "custom" view. Can this still be done?
>
>That's unimplimented so far. The issue is you also want to be able to set the selected tab and it would be great to do that without an index. I'm open to other ideas to support devs setting the tab (this is new functionality).

I've decided to actually expose the selected tab index as a property. The ColorViewTab will be indexed to follow the default tab ordering as well so you can do `SelectedIndex = (int)ColorViewTab.Palette` for example. This seems like the best overall solution and will allow adding more tabs/subviews/etc. in the future. You just won't be able to directly use the ColorViewTab as an index if you change the ordering.

timunie commented:
> I've decided to actually expose the selected tab index as a property. The ColorViewTab will be indexed to follow the default tab ordering as well so you can do SelectedIndex = (int)ColorViewTab.Palette for example. This seems like the best overall solution and will allow adding more tabs/subviews/etc. in the future. You just won't be able to directly use the ColorViewTab as an index if you change the ordering.

I agree.

robloo commented on src/Avalonia.Controls.ColorPicker/Themes/Fluent/ColorPicker.xaml:
```diff
@@ -0,0 +1,82 @@
+                - CustomPaletteColors
+              -->
+              <ColorView x:Name="FlyoutColorView"
+                         Color="{Binding $parent[ColorPicker].Color, Mode=TwoWay}"
```
@maxkatz6 Binding here with `Color` and `HsvColor` below works a little, but has a lot of issues. When selecting a color in the ColorView the value jumps around and is incorrect half the time. I need ideas for how to fix this, work around it, or if this is really an issue in the binding system. I'm not sure where to go from here.

Also note that binding to the TemplatedParent doesn't seem to work either which is why I went with the `$parent[ColorPicker].Color` syntax.

maxkatz6 replied:
I did not have a deep look, but I see a possible problem:

0. Previous **slider** HSV value is ```hsva(48, 0.6701030927835051, 0.3803921568627451, 1)``` and RGB - ```#ff615420```.
1. User [changes HSV property](https://github.com/AvaloniaUI/Avalonia/blob/bd8f7446d5136216919e0a59468fc1f0e83bd585/src/Avalonia.Controls.ColorPicker/ColorSlider/ColorSlider.cs#L412) of the **slider**. New value is hsva(48, 0.6701030927835051, 0.35163934426229515, 1). This value is sent with a binding notification to the **view**.
2. **View** [receives HSV and changes RGB](https://github.com/AvaloniaUI/Avalonia/blob/bd8f7446d5136216919e0a59468fc1f0e83bd585/src/Avalonia.Controls.ColorPicker/ColorView/ColorView.cs#L117) to ```#ff5a4e1e``` and sends binding notification to the **picker**. **disableUpdates** is true on the **view**, but false on the **picker**.
3. **Picker** [receives RGB and changes HSV](https://github.com/AvaloniaUI/Avalonia/blob/bd8f7446d5136216919e0a59468fc1f0e83bd585/src/Avalonia.Controls.ColorPicker/ColorView/ColorView.cs#L104) to ```hsva(48.00000000000001, 0.6666666666666666, 0.3529411764705882, 1)```. **disableUpdates** is true on the **view**, and true on the **picker**.
4. At this point ColorPicker.Hsv propagates through bindings code, hard to tell what's going on from the stacktrace.
5. Finally **Spectrum** receives HSV value, that was calculated on step number 3 inside of ColorPicker. Not the original one. And spectrum converts it to the RGB once more.

maxkatz6 replied:
As you can see, color from the slider goes quite a long way. And gets recalculated couple of times on the way.
**disableUpdates** field is useless when value goes from View to the Picker.

maxkatz6 replied:
Probably it would be easier to add some syncing code inside of the ColorPicker for (Hsv)Color and CustomPaletteColors props. As I don't see any obvious problem in bindings that can be fixed.

robloo replied:
Yes, this definitely looks like a circular property changed issue as you describe. I thought I designed around it but missed something.

For 0-3, I agree, that is how it should work. Note that ONLY the ColorView is responsible for the synchronization with RBG/HSV and Color/HsvColor properties. This is because all primitives are bound only through the HsvColor property. If a primitive needs RGB color it calculates it itself internally and separately. This means we should never have to worry about both Color and HsvColor binding changes occurring in a bad order or repeatedly updating each other. This is known to work just fine because the ColorView functions perfectly. So we know the issue is with ColorPicker.

Now, here is my original thinking that I realize is now incorrect: (thanks for coaxing in the right direction)

In the ColorPicker, the Color and HsvColor are bound directly to the ColorView. So if the ColorView is correct -- and it is -- the ColorPicker should just automatically get the values. Note that the ColorPicker has zero code-behind so I was stupidly thinking it shouldn't be propagating any changes back to the ColorView or doing RGB<->HSV sync. Even if it was, the ColorView shouldn't do anything because the values would be equal to the one that triggered the binding update to begin with. This was likely my mistake in hindsight.

Of course ColorPicker is derived from ColorView... so they both have the color sync logic running simultaneously. That was untested and was never intended. This seems to be the core problem. I'll have to add a way to disable the RGB<->HSV color sync logic internally in derived controls like ColorPicker or make this logic behave correctly with derived controls.

Anyway, thanks for pointing me in the right direction (calling out disableUpdates being different made me realize). I'll debug this later with lots of sleep.

robloo replied:
It seems that wasn't it. I also changed to normal bindings and they don't work at all. I think standard bindings can't reach across the Flyout/Popup to the root control. Even More, the control becomes unresponsive after the flyout closes.

robloo replied:
@maxkatz6 I confirmed two things:

1. The 'freezing' of the control that was happening when the flyout closes is now fixed with latest master. Not sure what that was
2. I rebuilt the template and put the ColorView outside of the Flyout just to test the bindings -- it worked. This proves the issue is that TemplateBinding in Avalonia does NOT work across a Popup/Flyout. I'll need help to fix this.

robloo replied:
#8367 was fixed resolving this conversation

robloo commented on src/Avalonia.Controls.ColorPicker/Themes/Fluent/ColorPicker.xaml:
```diff
@@ -0,0 +1,82 @@
+          <DropDownButton.Flyout>
+            <Flyout FlyoutPresenterClasses="NoPadding">
+              <!-- Skip the following:
+                - CustomPaletteColors
```
I also can't bind to CustomPaletteColors since it's a direct property and a read-only collection. I really don't want to sync the two collections manually with code-behind and event handling. However, I don't see another way. Am I missing something? Is it possible to bind in this scenario?

Note that if I allowed setting the collection it could work. But it's considered bad practice to allow setters for collecitons. Perhaps make it a get-only CLR property but a StyledProperty IEnumerable or something?

maxkatz6 replied:
Yes, as you said it's a readonly property (both CLR and DirectProp definition), and XAML doesn't know how to bind it. It doesn't know that ObservableCollection can do it.

I don't know other way, but use some help from code behind. Not exactly the same situation, but we did it for Flyout.PresenterClasses too - https://github.com/AvaloniaUI/Avalonia/blob/e51851e68f8ed93544713f428775113998a45a4e/src/Avalonia.Controls/Flyouts/FlyoutBase.cs#L564

maxkatz6 replied:
Is this kind of bindings supported in WPF/UWP? I don't remember if so, but I wish it could.

robloo replied:
> Is this kind of bindings supported in WPF/UWP? I don't remember if so, but I wish it could.

In UWP this certainly isn't supported. In WPF I also don't think so and have never seen it done. It's the one common case that has just never been handled correctly by any dependency property system...

Ideally, the DP system needs to special-case collections. A supported collection type could implement a new IApplyTo interface or something that would allow copying items from one collection into another by reference. While the interface would be easy I imagine the property system updates are not. This also hasn't been fully thought out so who knows what other issues there are. Anyway, I found something else.

> Not exactly the same situation, but we did it for Flyout.PresenterClasses too -

Thanks for the example.

I also found some cases where IEnumerable is used that seems to work-around this issue (taken from ItemsControl). I haven't seen done before in other XAML frameworks. Perhaps this is how it should be done in Avalonia?

https://github.com/AvaloniaUI/Avalonia/blob/56ef3adf065aba960648bb85466d03651583bce8/src/Avalonia.Controls/ItemsControl.cs#L105-L109

 (note it is still bad convention to have a setter for a collection/enumerable... but maybe that's the lesser of two evils to get this to work in the property system as-is)

maxkatz6 replied:
Now I think that public setter for a collections is not that uncommon in avalonia. Transitions is the same.

robloo replied:
If that is what is done in Avalonia I'll do it to get this to work. It just seems strange to me based on other XAML experience.

I still consider this an unsolved problem in the binding system though and it should be special cased.

robloo replied:
Switched to a StyledProperty following `ItemsControl`.

```csharp
        public static readonly StyledProperty<IEnumerable<Color>?> PaletteColorsProperty =
            AvaloniaProperty.Register<ColorView, IEnumerable<Color>?>(
                nameof(PaletteColors),
                null);

        public IEnumerable<Color>? PaletteColors
        {
            get => GetValue(PaletteColorsProperty);
            set => SetValue(PaletteColorsProperty, value);
        }
```

robloo committed:
Implement tab selection validation and automatic width

robloo committed:
Merge branch 'master' into colorpicker2

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0020883-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

robloo committed:
Support ignoring property changes in derived Color controls

robloo committed:
Use more standard binding in ColorPicker

This doesn't currently work for some reason

robloo committed:
Small improvements

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0020925-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

robloo committed:
Implement SelectedTabIndex in ColorView

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0020926-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

robloo committed:
Rename 'EnumToBooleanConverter' to 'EnumToBoolConverter'

Bool is shorter and is also slightly more commonly used in other converters. Bool matches the type as written in C# as well.

robloo committed:
Add FlatColorPalette

robloo committed:
Add SixteenColorPalette

robloo commented on src/Avalonia.Controls.ColorPicker/Themes/Fluent/ColorView.xaml:
```diff
@@ -0,0 +1,609 @@
+                      Grid.Row="0"
+                      Height="338"
+                      Width="350"
+                      SelectedIndex="{Binding SelectedTabIndex, RelativeSource={RelativeSource TemplatedParent}, Mode=TwoWay}">
```
Note here that if externally the SelectedTabIndex is set to 4 or something (outside the valid range) the TabControl will ignore it and replace it with a valid SelectedIndex. This however is not propogated back through the TwoWay binding so the SelectedTabIndex and SelectedIndex become out of sync.

The work-around for this is done in code-behind since I already have a ValidateSelectedTab() method. However, I'm not sure if the binding system should support this (UWP always had issues like this, WPF perhaps not).

robloo replied:
I just documented this in code, marking resolved.

robloo committed:
Add comment explaining rounding value selection

robloo committed:
Make ColorPalettes directory plural

robloo committed:
Add resources to control accent color section size in ColorPreviewer

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0020932-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

robloo committed:
Several fixes/improvements for ColorPreviewer

robloo committed:
Small optimization to read property only when needed

robloo commented on src/Avalonia.Controls.ColorPicker/ColorView/ColorViewTab.cs:
```diff
@@ -0,0 +1,26 @@
+    /// <remarks>
+    /// This is indexed to match the default control template ordering.
+    /// </remarks>
+    public enum ColorViewTab
```
Terminology for "Tab" in this enum and in the SelectedTabIndex should be mentioned/discussed. I'm attempting to make this control template-agnostic so it can be fully re-templated however one would like. However, these two areas need a word and I chose "Tab". Those that re-template this control may not have tabs at all but would still like some form of "subviews" (group boxes, etc.). So the term here may not be good.

While the term "Tab" is used in the API, in comments "subview/tab" is used which is a bit more generic. All terms I can think of are "Tabs", "Panels", "Subviews".

Does anyone have comments or concerns here?

timunie replied:
I want to just throw in "Page" here. It's really hard to find the right word, so I have no preference here.

maxkatz6 replied:
Either Tab or Page - we use both terms in the project. TabControl, TabStrip, but IPageTransition.
I don't think we use Subviews anywhere, except from UIKit/AppKit related code (terminology used by Apple).

robloo replied:
Subview feels like the correct term here even if it isn't used much in .NET or Avalonia. All the other terms are overly specific and won't necessary match with how the control is re-templated in the future. So in comments I'm using "tab/panel/page (subview)" but in the API "Tab" is removed as mentioned above everywhere but the enum.

robloo committed:
Merge branch 'master' into colorpicker2

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0020941-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0020942-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

robloo committed:
Merge branch 'master' into colorpicker2

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0021027-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

robloo committed:
Remove unused field to avoid warnings

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0021124-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

robloo committed:
Set the CustomPalette property in XAML

robloo committed:
Remove 'Tab' terminology from SelectedIndex property

robloo committed:
Merge branch 'master' into colorpicker2

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0021254-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

robloo committed:
Remove 'Custom' term from Palette properties

robloo committed:
Allow setting the PaletteColors collection as an IEnumerable

In WPF it was standard to have read-only collections. However, it's necessary to bind this collection between the ColorPicker and ColorView so it cannot be read-only. ItemsControl in Avalonia implements the Items property a similar way.

robloo committed:
Partial work-around for Flyout binding issues

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0021278-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

robloo committed:
Merge branch 'master' into colorpicker2

robloo committed:
Add IsColorModelVisible and IsComponentSliderVisible properties

robloo committed:
Use nested styles in ColorView

robloo committed:
Implement IsAlphaEnabled using coercion

robloo committed:
Only set slider IsVisible once

robloo committed:
Only disable alpha editing controls when IsAlphaEnabled is false

Rely on IsAlphaVisible for visibility

robloo committed:
Remove unused ValueConverterGroup

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0021395-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

robloo committed:
Merge branch 'master' into colorpicker2

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0021411-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

robloo committed:
Merge branch 'master' into colorpicker2

robloo committed:
Switch to TemplateBinding in ColorPicker where possible

robloo committed:
Standardize ColorControl resource names

This follows the Fluent standard where the resource is prefixed with the control itself

robloo committed:
Better implement ColorView backgrounds in the template for easier re-styling

robloo committed:
Add code to hide the tab strip when only one tab is visible

robloo commented:
@maxkatz6 I'm marking this as ready -- it is complete enough for the first preview.

There are a few topics left open but can be continued later as needed (or if the PR is still open):
 1. The only main thing I'm still debating is changing some of the property names. "IsColor..." to just "Is...". The term `Color` I think is unnecessary and inferred from context being on the `ColorView` itself. I will compare with WinUI naming as well before making the final decision. Any thoughts you might have on this will also be considered.
 2. As a secondary item, I would like to make the `ColorPicker` support showing just a palette without having a tab strip shown. (so it would basically just show a flyout of pre-defined colors to select in the grid) I ran into a problem hiding the background border (without doing everything in code-behind) therefore this stalled for now. However, my goal is to support this as well.
 3. Finally, I may add a new `ColorView` template that closely matches the WinUI horizontal layout of the ColorPicker. That would essentially allow usage of this as a drop-in replacement (from a design standpoint) for apps being ported from WinUI.

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0021629-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

maxkatz6 commented:
@robloo 

1. `IsColorPaletteVisible` can be renamed to `IsPaletteVisible`. But `IsModelVisible` just doesn't sound right. Would be easier to keep it as it is, I guess. 
2. Yeah, I don't see an easy way to do it. Unless you remove "TabBackgroundBorder" and instead set background color to a part of the TabControl template itself. So far ColorView has default TabControl style.
3. I don't think it's necessary.

maxkatz6 commented on src/Avalonia.Controls.ColorPicker/Themes/Fluent/ColorView.xaml:
```diff
@@ -0,0 +1,643 @@
+                    </Style>
+
+                    <Style Selector="^:selected /template/ Rectangle#BorderRectangle">
+                      <Setter Property="Stroke" Value="{Binding Converter={StaticResource ContrastBrush}, ConverterParameter={DynamicResource TextControlForeground}}" />
```
ConverterParameter doesn't accept bindings (including DynamicResource) unfortunately. Inside of the converter it will be just a InstancedBinding object instead of actual resource. Multibinding can be used instead.

maxkatz6 replied:
Unless https://github.com/AvaloniaUI/Avalonia/pull/7624

robloo replied:
I know bindings don't work until #7624; however, this code always seemed to work. That parameter is used if alpha threshold is below 50%. If you are sure, I will retest.

robloo replied:
You are of course correct here, the binding itself is being passed as the parameter.

I was thinking of the Slider which does need to do this type of logic (and it is working); however, it's done in a completely different way so has no issue.

For the Palette here its extremely rare for a color (defined in a palette) to actually have any transparency let alone over 50%. So I have just disabled that logic by removing the parameter itself. If ever Avalonia supports this it can be re-enabled. Instructions and comments are added to the code.

robloo commented on src/Avalonia.Controls.ColorPicker/Themes/Fluent/ColorPicker.xaml:
```diff
@@ -0,0 +1,92 @@
+        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
+        xmlns:local="using:Avalonia.Controls"
+        xmlns:converters="using:Avalonia.Controls.Converters"
+        x:CompileBindings="False">
```
Note that I cannot use compiled bindings here as it throws errors. I didn't investigate this much.

maxkatz6 replied:
TargetType should be set on ControlTemplate.
And DataType should be set on DataTemplate.
In some cases, compiler can find out about type from the context, but not sure about current situation. At least Rider is confused without these types, but Rider has its own analyzers.

It seems there is another problem with Avalonia Xaml compiler:
```
  A binding with a TemplatedParent RelativeSource has to be in a ControlTemplate. Line 120, position 23. 
```

robloo replied:
Yea, I know about setting the types but there were additional errors like the one you mentioned so I didn't bother looking at this further.

maxkatz6 replied:
I think I found the issue. Will try to fix it later.

robloo replied:
I added several x:Type declarations just in case you are able to fix the compiler. Otherwise, I'm also fine with leaving it as-is.

robloo replied:
This was fixed with #8466, Thanks!

robloo commented on src/Avalonia.Controls.ColorPicker/Themes/Fluent/ColorView.xaml:
```diff
@@ -0,0 +1,643 @@
+        xmlns:primitives="using:Avalonia.Controls.Primitives"
+        xmlns:pc="clr-namespace:Avalonia.Controls.Primitives.Converters;assembly=Avalonia.Controls.ColorPicker"
+        xmlns:globalization="clr-namespace:System.Globalization;assembly=mscorlib"
+        x:CompileBindings="False">
```
Note that I cannot use compiled bindings here as it throws errors. I didn't investigate this much.

robloo replied:
This mostly worked after #8466 I only had to add the x:DataType to the last nested style that uses the binding (the same one discussed above with the converter parameter). I didn't know that x:DataType could be added to Style itself actually.

robloo commented on src/Avalonia.Controls.ColorPicker/Themes/Fluent/ColorView.xaml:
```diff
@@ -0,0 +1,643 @@
+    <pc:ThirdComponentConverter x:Key="ThirdComponent" />
+    <converters:ColorToDisplayNameConverter x:Key="ColorToDisplayName" />
+    <converters:ColorToHexConverter x:Key="ColorToHex" />
+    <converters:EnumToBoolConverter x:Key="EnumToBool" />
+    <converters:CornerRadiusFilterConverter x:Key="TopCornerRadiusFilterConverter" Filter="TopLeft, TopRight"/>
+    <converters:CornerRadiusFilterConverter x:Key="BottomCornerRadiusFilterConverter" Filter="BottomLeft, BottomRight"/>
```
I think I'm going to pull all the common converters out into a shared file (Fluent.xaml or Default.xaml). There is no reason to declare this in each file separately.

maxkatz6 replied:
You also can define them as static properties and reference with `{x:Static}`.
I.e.
```
public class CornerRadiusFilterConverter : IValueConverter
{
    public static CornerRadiusFilterConverter Instance { get; } = new();
   
    public object? Convert()...
}
```

robloo replied:
Hm, that's an interesting idea to consider as a new convention. For now will probably keep them in XAML though as that is pretty standard.

robloo replied:
Pulled out all converters and standardized naming. I see you are also doing the static instance with `BoolConverters.And` for instance.

robloo commented on src/Avalonia.Controls.ColorPicker/Themes/Fluent/ColorView.xaml:
```diff
@@ -0,0 +1,643 @@
+                  <Style Selector="ListBoxItem">
+                    <Setter Property="Background" Value="Transparent" />
+                    <Setter Property="Padding" Value="0" />
+                    <Setter Property="Template">
```
The template here probably should be pulled out as a property on ColorView itself. That would allow changing to elliptical color swatches or something without re-templating.

maxkatz6 replied:
I think for now it's flexible enough as it is. User can override ListBoxItem.Template with nested selectors.

robloo replied:
Ok, well I won't add the property then for now and will just keep it in mind.

robloo commented on src/Avalonia.Controls.ColorPicker/ColorView/ColorView.Properties.cs:
```diff
@@ -0,0 +1,495 @@
+        /// <summary>
+        /// Defines the <see cref="ShowAccentColors"/> property.
+        /// </summary>
+        public static readonly StyledProperty<bool> ShowAccentColorsProperty =
```
Note: All other bool properties here use "Is" as the prefix. "Show" should be reconsidered here although it is common as an alternative in .NET and does shorten length and make more readable in cases like this.

robloo replied:
Tentatively renamed to `sAccentColorsVisible`. Full overview will be in a comment.

robloo committed:
Standardize converters

robloo committed:
Remove unused code

robloo committed:
Specify TargetType and DataType in most templates

robloo committed:
Add missing properties to ColorPicker template

robloo commented:
@maxkatz6 A full review was done of the bool properties. Since this is finalizing the API and different from WinUI a bit more time was spent on this. Please let me know if you (or anyone else reading this) agree or have questions/concerns.

The main column to look at is the "New" one on the right. Some alternative ideas are listed to the immediate right of the "New" column but were decided against.

![image](https://user-images.githubusercontent.com/17993847/177899981-b7059347-1070-4e70-8baa-4d692cd9fb37.png)

1. "Section" could be added to better differentiate the tab visibility properties. "Section" is more generic than "Tab" but still may not be relevant in future re-templated controls.
2. "IsComponentTextInputVisible" could be shortened to "IsComponentInputVisible". This matches better with "IsHexInputVisible"; however, may then be confused with "IsComponentSliderVisible" which is also another input. 
3. Both "IsColorModelVisible" and "IsColorPreviewVisible" need to keep the term "Color".
4. Both "IsComponentsVisible" and "IsAccentColorsVisible" use the plural term which sounds a bit strange and may be grammatically incorrect. However, this is sometimes done in .NET to follow the "Is" convention.

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0021737-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

robloo committed:
Merge branch 'master' into colorpicker2

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0021773-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

robloo committed:
Enable compiled bindings for ColorPicker and ColorView

robloo commented on src/Avalonia.Controls.ColorPicker/Themes/Fluent/ColorPicker.xaml:
```diff
@@ -0,0 +1,91 @@
+    </Setter>
+  </Style>
+
+  <!-- Adjust Background within Flyout -->
+  <!-- Note: This is implemented but there seems to be an issue and the selector can't match across the Flyout -->
+  <Style Selector="ColorPicker /template/ ColorView#FlyoutColorView /template/ Border#TabBackgroundBorder">
+    <Setter Property="CornerRadius" Value="{DynamicResource TopOverlayCornerRadius}" />
+  </Style>
```
Also note this doesn't work for some reason. I'm not sure if the comment is correct "This is implemented but there seems to be an issue and the selector can't match across the Flyout" or I made a mistake in the Selector (examples like this aren't in the docs but I did see some usage like this in other code).

maxkatz6 replied:
Selector by itself it valid. 

Similar limitation to this one https://docs.avaloniaui.net/docs/styling/troubleshooting#missing-style-pseudoclass-trigger-selector

Basically, TemplateBinding priority is higher than Style-Setter priority, but lower than Style-Trigger-Setter (with a pseudoclass)
https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Base/Data/BindingPriority.cs#L18-L37

maxkatz6 replied:
I would recommend leaving it as it is. It should be easier to work around it with control themes later.

robloo replied:
Yes, agree. I will have to rework the styles in control themes anyway so this will all be gone over again. 

Thanks for explaining how the priority is causing this situation.

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0021775-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

robloo committed:
Merge branch 'master' into colorpicker2

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0021785-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

maxkatz6 commented:
Personally, I don't have anything against of current naming in this PR, seems reasonable.
Details like `Border#TabBackgroundBorder` corner radius can be revisited later with ControlThemes.

robloo commented:
Well, I thought about the property naming some more. The name of the primitive is ColorSpectrum and the interface IColorPalette. So yes, it makes sense to keep the existing color term in the name for those two. Then to be consistent, ColorComponents should be used for the third tab. So I might end up only changing ShowAccentColors to IsAccentColorsVisible (all others use "visible" and "is" terminology).

I found another thing where the color previewed drop shadow should really be removed when accent colors are not shown.

If you could pause auto-merge I'll try to make the changes within the next few days.

maxkatz6 commented:
@robloo paused

robloo committed:
Merge branch 'master' into colorpicker2

robloo committed:
Rename `ShowAccentColors` to `IsAccentColorsVisible` and disable drop shadow when false

robloo commented:
@maxkatz6 Thanks, updates are made and this should be all set for now.

I will make the next round of updates converting to `ControlTheme` after those PRs are finalized and merged. Although, it still isn't clear to me how to really extend FluentTheme at that point.

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0022103-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

robloo commented:
Not sure what conflicts it is talking about unless it means the IntegrationTests failures. I merged in master branch with no issue.

I'm assuming the IntegrationTests are still a WIP? Let me know how to proceed here.

maxkatz6 commented:
@robloo framework used for integration tests is very unstable, and it fails quite frequently. It's not yet solved, but PRs can be merged anyway, as it doesn't block merging. I can do it now, if you want to continue with control themes in another PR.

robloo commented:
> I can do it now, if you want to continue with control themes in another PR.

Yes, that was my plan and I'm not sure when the other PRs will be completed. If you could merge this soon that would be great. Sorry for not being more clear about it.

timunie commented:
@robloo very beautiful and powerful color picker 👍 Can't wait to use it in v11.0 (I know I can use nightly builds, but they are breaking other things for me atm)

I think we now should create the needed docs for it. I can support you here if you want me to.

robloo commented:
@timunie Thanks! Hope you find it useful.

> I think we now should create the needed docs for it. I can support you here if you want me to.

I haven't forgotten about this and already started a draft of some ideas some time ago. However, there is no rush to complete it. As you know new docs are getting merged in right now. 

I will plan on finishing this up myself due to the nature of this control. Your review after the initial draft will be very useful though.
